crates/matrix-sdk-ui/src/authentication/oidc.rs

```diff
@@ -0,0 +1,214 @@
+use std::{
+    collections::{hash_map::DefaultHasher, HashMap},
+    fs,
+    fs::File,
+    hash::{Hash, Hasher},
+    io::{BufReader, BufWriter},
+    path::PathBuf,
+};
+
+use matrix_sdk::oidc::types::registration::VerifiedClientMetadata;
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, thiserror::Error)]
+pub enum OidcRegistrationsError {
+    #[error("Failed to use the supplied base path.")]
+    InvalidBasePath,
+}
+
+/// The data needed to restore an OpenID Connect session.
+#[derive(Debug)]
+pub struct OidcRegistrations {
+    /// The path of the file where the registrations are stored.
+    file_path: PathBuf,
+    /// The hash for the metadata used to register the client.
+    /// This is used to check if the client needs to be re-registered.
+    metadata_hash: u64,
+    /// Pre-configured registrations for use with issuers that don't support
+    /// dynamic client registration.
+    static_registrations: HashMap<String, String>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct RegistrationData {
```

Comment:
nit: please add a doc comment here too

And maybe we could make it obvious in the name that it's saved to disk, something like `FrozenRegistrationData`?